File: homing-workspace-shell/src/main/java/hue/captains/singapura/js/homing/workspace/shell/PersistenceAttacherModule.java
package hue.captains.singapura.js.homing.workspace.shell;

import hue.captains.singapura.js.homing.core.DomModule;
import hue.captains.singapura.js.homing.core.Exportable;
import hue.captains.singapura.js.homing.core.ExportsOf;
import hue.captains.singapura.js.homing.core.ImportsFor;
import hue.captains.singapura.js.homing.core.ModuleImports;
import hue.captains.singapura.js.homing.workspace.persistence.WorkspaceStatePersistenceModule;

import java.util.List;

/**
 * {@code PersistenceAttacher} — Phase 4 of the workspace-shell chrome.
 * Thin wrapper around {@code WorkspaceStatePersistence.attach}: pulls
 * {@code spec.kind} for the {@code workspaceKind} param, supplies
 * {@code window.localStorage} as the storage, returns the constructed
 * persistence layer (or {@code null} on failure, logged).
 *
 * <p>What this <i>does not</i> do — by intent in this phase:</p>
 *
 * <ul>
 *   <li>Does not call {@code tryRestore}. RFC 0030 P3 retired
 *       snapshot-based boot; restoration is event-replay (Phase 9).</li>
 *   <li>Does not call {@code create}. The persister is not used by V2
 *       yet — RFC 0030 P3 likewise retired the snapshot persister; only
 *       checkpoint writes (Phase 8) actually save state.</li>
 * </ul>
 *
 * <p>The attach is still useful: the returned layer's {@code store} is
 * what {@code ResetControls} (Phase 18) calls {@code .clear(kind)} on
 * to wipe any lingering legacy localStorage data. So Phase 4 effectively
 * makes the storage layer present and inspectable; Phases 8/9 do the
 * real save/restore via the checkpoint store.</p>
 *
 * <p>Explicit Substrate: instance + INSTANCE + no static methods.
 * {@code attach} is sync — the wrapped function is sync — so no
 * Promise plumbing.</p>
 *
 * @since post-RFC-0034 workspace chrome decomposition — Phase 4
 */
public record PersistenceAttacherModule()
        implements DomModule<PersistenceAttacherModule> {

    public static final PersistenceAttacherModule INSTANCE = new PersistenceAttacherModule();

    /** The {@code PersistenceAttacher} JS class. */
    public record PersistenceAttacher() implements Exportable._Class<PersistenceAttacherModule> {}

    @Override
    public ImportsFor<PersistenceAttacherModule> imports() {
        return ImportsFor.<PersistenceAttacherModule>builder()
                .add(new ModuleImports<>(
                        List.of(new WorkspaceStatePersistenceModule.WorkspaceStatePersistence()),
                        WorkspaceStatePersistenceModule.INSTANCE))
                .build();
    }

    @Override
    public ExportsOf<PersistenceAttacherModule> exports() {
        return new ExportsOf<>(INSTANCE, List.of(new PersistenceAttacher()));
    }
}

File: homing-workspace-shell/src/main/resources/homing/js/hue/captains/singapura/js/homing/workspace/shell/PersistenceAttacherModule.js
// =============================================================================
// PersistenceAttacher — Phase 4 of the workspace-shell chrome.
//
// Thin wrapper around WorkspaceStatePersistence.attach. Returns the
// constructed persistence layer (or null on failure, logged).
//
// Explicit Substrate doctrine:
//   - Instance methods on INSTANCE singleton (no static).
//   - attach is SYNC.
//   - ALL collaborators in constructor: _persistence.
//
// Instance fields (graph-traversable):
//   _persistence : WorkspaceStatePersistence — defaults to the global
// =============================================================================

class PersistenceAttacher {

    constructor(deps) {
        deps = deps || {};
        this._persistence = deps.persistence || WorkspaceStatePersistence;
    }

    /**
     * Attach the persistence layer for a workspace kind. Returns the
     * layer (whose .store + .workspaceKind props serve later phases),
     * or null if attach throws (logged).
     *
     * Per-call data only: workspaceKind, storage. Collaborator
     * (_persistence) is fixed at construct time.
     */
    attach(opts) {
        if (!opts || !opts.workspaceKind) {
            throw new Error('[PersistenceAttacher] opts.workspaceKind required');
        }
        const storage = (opts.storage !== undefined)
                      ? opts.storage
                      : (typeof window !== 'undefined' && window.localStorage)
                            ? window.localStorage
                            : null;
        if (!storage) {
            console.warn('[PersistenceAttacher] no storage available '
                       + '(no window.localStorage, none supplied) — skipping attach.');
            return null;
        }
        try {
            return this._persistence.attach({
                workspaceKind: opts.workspaceKind,
                storage:       storage,
                paramsCodecs:  {}    // codecs registered separately by CodecRegistrar
            });
        } catch (err) {
            console.error('[PersistenceAttacher] attach failed for kind "'
                        + opts.workspaceKind + '":', err);
            return null;
        }
    }
}

PersistenceAttacher.INSTANCE = new PersistenceAttacher();
